static/js/pm_list.js

```diff
@@ -1,107 +1,189 @@
+        const conversations = pm_list_data.get_conversations();
+        const active_conversation = conversations.find((conversation) => conversation.is_active);
+
+        if (active_conversation) {
+            const node = [pm_list_dom.keyed_pm_li(active_conversation)];
+            const new_dom = pm_list_dom.pm_ul(node);
+            set_dom_to(new_dom);
+        } else {
+            // Otherwise, empty the section.
+            $(".expanded_private_messages").empty();
+            prior_dom = undefined;
         }
-
-        vdom.update(replace_content, find, new_dom, prior_dom);
-        prior_dom = new_dom;
+    } else {
+        const new_dom = _build_private_messages_list();
+        set_dom_to(new_dom);
     }
 }
 
 export function expand() {
-    private_messages_open = true;
+    private_messages_collapsed = false;
+    $("#toggle-private-messages-section-icon").addClass("fa-caret-down");
+    $("#toggle-private-messages-section-icon").removeClass("fa-caret-right");
     update_private_messages();
-    if (pm_list_data.is_all_privates()) {
-        $(".top_left_private_messages").addClass("active-filter");
-    }
 }
 
 export function update_dom_with_unread_counts(counts) {
+    // In theory, we could support passing the counts object through
+    // to pm_list_data, rather than fetching it directly there. But
+    // it's not an important optimization, because it's unlikely a
+    // user would have 10,000s of unread PMs where it could matter.
     update_private_messages();
+    // This is just the global unread count.
     set_count(counts.private_message_count);
 }
 
-function should_expand_pm_list(filter) {
-    const op_is = filter.operands("is");
-
-    if (op_is.length >= 1 && op_is.includes("private")) {
-        return true;
-    }
+export function highlight_all_private_messages_view() {
+    $("#private_messages_list").addClass("active_private_messages_list");
+    $("#more-private-conversations-sidebar-title").addClass(
+        "active-more-private-conversations-sidebar-title",
+    );
+}
 
-    const op_pm = filter.operands("pm-with");
+function unhighlight_all_private_messages_view() {
+    $("#private_messages_list").removeClass("active_private_messages_list");
+    $("#more-private-conversations-sidebar-title").removeClass(
+        "active-more-private-conversations-sidebar-title",
+    );
+}
 
-    if (op_pm.length !== 1) {
-        return false;
+export function handle_narrow_activated(filter) {
+    const active_filter = filter;
+    const is_all_private_message_view = _.isEqual(active_filter.sorted_term_types(), [
+        "is-private",
+    ]);
+    const narrow_to_private_messages_section = active_filter.operands("pm-with").length !== 0;
+
+    if (is_all_private_message_view) {
+        highlight_all_private_messages_view();
+    } else {
+        unhighlight_all_private_messages_view();
     }
+    if (narrow_to_private_messages_section) {
+        update_private_messages();
```

Comment:
@timabbott , I've just made this 1 change in your updates made to the code, to ensure that we do not call `update_private_messages` unnecessarily if we're not narrowing inside any PMs or the `All Private Messages` view, similar to how we handle `handle_narrow_activated` in `top_left_corner` and `stream_list` to just highlight and unhighlight the section on call;
 
 As running `update_private_messages` every time unnecessarily would eventually lead to node test failure for `narrow_activate`;

Reply:
OK; I'll take a look. I think I was concerned that the logic here was wrong in certain corner cases where you're switching away from one narrow to another (where one is a specific PM thread and the other is not).

Reply:
Cool, but I think the this new function works perfect in all cases of narrowing (I tested it locally).